simplic-corelib/Simplic.CoreLib/Collections/Generic/StatefulCollection/StatefulCollection.cs

```diff
@@ -1,275 +1,409 @@
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item,
+                    ReaddFromRemovedItems = true
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
+                {
+                    newItems.Add(item);
+                    // Remove item from the list of removed items
+                    items.Add(item);
+                    removedItems.Remove(item);
+
+                    var itemAdded = new ItemAddedEventArgs
+                    {
+                        Item = item
+                    };
+                    ItemAdded?.Invoke(this, itemAdded);
+                }
+            }
+            else if (!items.Contains(item) && !newItems.Contains(item))
+            {
+                var args = new AddItemEventArgs
+                {
+                    Item = item
+                };
+                AddItem?.Invoke(this, args);
+
+                if (!args.Cancel)
+                {
+                    newItems.Add(item);
+
+                    var itemAdded = new ItemAddedEventArgs
+                    {
+                        Item = item
+                    };
+                    ItemAdded?.Invoke(this, itemAdded);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add a range of items. Each item will be handled as described for <see cref="Add(T)"/>
+        /// </summary>
+        /// <param name="items">List of items to handle</param>
+        public void AddRange(IEnumerable<T> collection)
+        {
+            foreach (var item in collection)
+            {
+                Add(item);
+            }
+        }
+
+        /// <summary>
+        /// Clear the howle collection
+        /// </summary>
+        public void Clear()
+        {
+            var args = new ClearCollectionEventArgs();
+            ClearCollection?.Invoke(this, args);
+
+            if (!args.Cancel)
+            {
+                items.Clear();
+                newItems.Clear();
+                removedItems.Clear();
+
+                var collectionClearedArgs = new CollectionClearedEventArgs();
+
+                CollectionCleared?.Invoke(this, collectionClearedArgs);
+            }
+        }
+
+        /// <summary>
+        /// Check whether an items exists in the items list or the new items list
+        /// </summary>
+        /// <param name="item">Item to proof</param>
+        /// <returns>True if the item exists</returns>
+        public bool Contains(T item)
+        {
+            if (items.Contains(item) || newItems.Contains(item))
+                return true;
+
+            return false;
+        }
+
+        /// <summary>
+        /// Copy current items into an array
+        /// </summary>
+        /// <param name="array">Array to copy items to</param>
+        /// <param name="arrayIndex">Start index</param>
+        public void CopyTo(T[] array, int arrayIndex)
```

Comment:
I think a method such as `ToArray()` is way better. but i assume this is a legacy method

Reply:
possible index out of range exception. you should check the `arrayIndex` if its inside the arrays size. To be honest, this method does not belong here. its not this objects job to make sure array is not empty, index is within the range and other possible bugs. Just return a new array